File: linked_list/singly.py
class Node:
    def __init__(self, value:int):
        self.value = value
        self.next = None

class SinglyLL:
    def __init__(self):
        self.head = None

    def add(self, value: int):
        new_node = Node(value)

        if self.head == None:
            self.head = new_node
            return

        curr = self.head
        while curr.next:
            curr = curr.next

        curr.next = new_node
        return

    def insert(self, index: int, value: int) -> None:
        new_node = Node(value)

        if index == 0:
            new_node.next = self.head
            self.head = new_node
            return

        curr_idx = 0
        curr = self.head
        while curr and curr_idx < index - 1:
            curr = curr.next
            curr_idx += 1

        if curr is None:
            print(f"{index}th exist nahi karti")

        new_node.next = curr.next
        curr.next = new_node
        return


    def remove(self, index: int) -> None:
        curr_idx = 0
        curr = self.head
        while curr_idx != (index-1):
            if curr.next:
                curr = curr.next
                curr_idx += 1
            else:
                print(f"LinkedList has size of {curr_idx+1} can't remove at {index} index")
                return

        curr.next = curr.next.next
        return

    def pop(self) -> int|None:
        if self.head is None:
            print("list empty hai")
            return None

        if self.head.next is None:
            value = self.head.value
            self.head = None
            return value

        curr = self.head
        while curr.next.next:
            curr = curr.next

        value = curr.next.value
        curr.next = None
        return value

    def get(self, index: int) -> int|str|None:
        if self.head is None:
            return "list empty hai"
        curr_idx = 0
        curr = self.head
        while curr and curr_idx < index:
            curr = curr.next
            curr_idx += 1

        if curr is None:
            return f"{index}th exist nahi karti"

        return curr.value 

    def search(self, value: int) -> str:
        curr_idx = 0
        curr = self.head
        while curr is not None and curr.value != value:
            curr_idx += 1
            curr = curr.next

        if curr is None:
            return f"koi node nahi hai with value: {value}"

        return f"yeh lo index: {curr_idx} aur uski value:{curr.value}"

    def __str__(self):
        values = []
        curr = self.head
        while curr:
            values.append(str(curr.value))
            curr = curr.next
        return "->".join(values) + "->end"

sll = SinglyLL()
sll.add(5)
sll.add(5)
sll.add(3)
sll.add(1)
sll.add(2)
sll.add(3)
sll.add(1)
sll.add(2)
sll.insert(index=1, value=4)
sll.remove(1)
sll.pop()
a = sll.get(0)
print(a)
b = sll.search(3)
print(b)
print(sll)

File: linked_list/singly.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}

type SinglyLL struct {
	head *Node
}

func NewSinglyLL() *SinglyLL {
	return &SinglyLL{}

}

func (sll *SinglyLL) AddNode(value int) {
	new_node := Node{
		data: value,
		next: nil,
	}

	if sll.head == nil {
		sll.head = &new_node
		return
	}

	curr := sll.head
	for curr.next != nil {
		curr = curr.next
	}

	curr.next = &new_node
	return
}

func (sll *SinglyLL) Insert(index int, value int) {
	new_node := Node{
		data: value,
		next: nil,
	}

	if index == 0 {
		new_node.next = sll.head
		sll.head = &new_node
		return
	}

	curr_idex := 0
	curr := sll.head
	for curr != nil && curr_idex < index-1 {
		curr = curr.next
		curr_idex += 1
	}

	if curr == nil {
		fmt.Printf("%dth exist nahi karta sirji\n", index)
	}

	new_node.next = curr.next
	curr.next = &new_node
	return
}

func (sll *SinglyLL) Remove(index int) {
	curr := sll.head
	curr_idex := 0

	for curr_idex != index-1 {
		if curr.next != nil {
			curr = curr.next
			curr_idex += 1
		} else {
			fmt.Printf("linkedlist ka size hi %d etna hai kaise %dth pr remove kareji mai??", curr_idex+1, index)
		}
	}

	curr.next = curr.next.next
	return
}

func (sll *SinglyLL) Pop() (int, error) {
	if sll.head == nil {
		return 0, fmt.Errorf("Linked list to empty hai boss!!")
	}

	if sll.head.next == nil {
		data := sll.head.data
		sll.head = nil
		return data, nil
	}

	curr := sll.head
	for curr.next.next != nil {
		curr = curr.next
	}

	data := curr.next.data
	curr.next = nil
	return data, nil
}

func (sll *SinglyLL) Get(index int) (int, error) {
	if sll.head == nil {
		return 0, fmt.Errorf("List to empty hai boss!!")
	}
	curr_idex := 0
	curr := sll.head

	for curr != nil && curr_idex < index {
		curr = curr.next
		curr_idex += 1
	}

	if curr == nil {
		return 0, fmt.Errorf("%dth index hi exist nhi karta!!", index)
	}
	return curr.data, nil
}

func (sll *SinglyLL) Search(data int) (string, error) {
	curr_idex := 0
	curr := sll.head
	for curr != nil && curr.data != data {
		curr_idex += 1
		curr = curr.next
	}

	if curr == nil {
		return "", fmt.Errorf("koi node nahi hai with data: %d", data)
	}

	return fmt.Sprintf("yeh lo index: %d aur uski value:%d", curr_idex, curr.data), nil
}

func (sll *SinglyLL) Print() {
	curr := sll.head
	for curr != nil {
		fmt.Printf("%d -> ", curr.data)
		curr = curr.next
	}
	fmt.Println("nil")
}

func main() {
	sll := NewSinglyLL()
	sll.AddNode(5)
	sll.AddNode(5)
	sll.AddNode(3)
	sll.AddNode(1)
	sll.AddNode(2)
	sll.AddNode(3)
	sll.AddNode(1)
	sll.AddNode(2)
	sll.Insert(1, 4)
	sll.Remove(1)
	val, err := sll.Pop()
	if err != nil {
		panic(err)
	}
	fmt.Printf("Pop Value: %d\n", val)
	val1, err := sll.Get(1)
	if err != nil {
		panic(err)
	}
	fmt.Printf("Get Value: %d\n", val1)
	val2, err := sll.Search(3)
	if err != nil {
		panic(err)
	}
	fmt.Printf("%s\n", val2)
	sll.Print()
}
